File: sj/shared/src/main/scala/ReadBinary.scala
object ReadBinary {
  def main(args: Array[String]): Unit = {
    println("Hello World")

    // defs
    val fileName = "/Users/vk/software/TestIO/cc/bin/test_binary.bin"
    val bufferSize = 1024 * 4
    val numBuffers = 1000 * 1000

    // java.io
      javaio(fileName, bufferSize, numBuffers) 

    println("Bye Bye World")
  }

  def javaio(fileName: String, bufferSize: Int, numBuffers: Int): Unit = {
    import java.io._
    val inputStream = new FileInputStream(fileName)
    val buffer = Array.fill[Byte](bufferSize)(0)

    for (i <- 0 until numBuffers) {
      inputStream.read(buffer)
    }

    inputStream.close
  }

  def randomacess(fileName: String, bufferSize: Int, numBuffers: Int): Unit = {
    import java.io._
    val inputStream = new FileInputStream(fileName)
    val buffer = Array.fill[Byte](bufferSize)(0)

    for (i <- 0 until numBuffers)
      inputStream.read(buffer)

    inputStream.close
  }
  
  def buffered(fileName: String, bufferSize: Int, numBuffers: Int): Unit = {
    import java.io._
    val inputStream = new FileInputStream(fileName)
    val buffer = Array.fill[Byte](bufferSize)(0)

    for (i <- 0 until numBuffers)
      inputStream.read(buffer)

    inputStream.close
  }

  def bytebuffer(fileName: String, bufferSize: Int, numBuffers: Int): Unit = {
    import java.io._
    val inputStream = new FileInputStream(fileName)
    val buffer = Array.fill[Byte](bufferSize)(0)

    for (i <- 0 until numBuffers)
      inputStream.read(buffer)

    inputStream.close
  }

  def memmapped(fileName: String, bufferSize: Int, numBuffers: Int): Unit = {
    import java.io._
    val inputStream = new FileInputStream(fileName)
    val buffer = Array.fill[Byte](bufferSize)(0)

    for (i <- 0 until numBuffers)
      inputStream.read(buffer)

    inputStream.close
  }
}

File: cc/src/read_binary.cpp
#include <iostream>
#include <fstream>
#include <stdio.h>

// define buffer size in bytes
#define BUFFER_SIZE 100

// define the number of buffers to be written
#define NUM_BUFFERS 1000 * 1000

int main(int argc, char** argv) {
    std::cout << "Hello World" << std::endl;

    std::ifstream file("test_binary.bin", std::ios::in | std::ios::binary);

    char buffer[BUFFER_SIZE];

    for (auto i=0; i<NUM_BUFFERS; i++) {
        file.read(buffer, BUFFER_SIZE);

        /*
        for (auto j=0; j<BUFFER_SIZE; j++)
            printf("%d", buffer[j]);
        std::cout << std::endl;
        */
    }

    file.close();

    std::cout << "Bye Bye World" << std::endl;
}
